src/systems/joint_state_publisher/JointStatePublisher.cc

```diff
@@ -51,33 +51,68 @@ void JointStatePublisher::Configure(
     return;
   }
 
-  // Create the position, velocity, and force components for the joint.
-  std::vector<Entity> childJoints = _ecm.ChildrenByComponents(
-      this->model.Entity(), components::Joint());
-  for (const Entity &joint : childJoints)
+  // If a joint_name is specified in the plugin, then only publish the
+  // specified joints. Otherwise, publis all the joints.
+  if (_sdf->HasElement("joint_name"))
   {
-    // Create joint position component if one doesn't exist
-    if (!_ecm.EntityHasComponentType(joint,
-          components::JointPosition().TypeId()))
+    sdf::Element *ptr = const_cast<sdf::Element *>(_sdf.get());
+    sdf::ElementPtr elem = ptr->GetElement("joint_name");
+    while (elem)
     {
-      _ecm.CreateComponent(joint, components::JointPosition());
-    }
+      std::string jointName = elem->Get<std::string>();
+      gazebo::Entity jointEntity = this->model.JointByName(_ecm, jointName);
+      if (jointEntity != kNullEntity)
+      {
+        this->CreateComponents(_ecm, jointEntity);
+      }
+      else
+      {
+        ignerr << "Joint with name[" << jointName << "] not found. "
+          << "The JointStatePublisher will not publishe this joint.\n";
+      }
 
-    // Create joint velocity component if one doesn't exist
-    if (!_ecm.EntityHasComponentType(joint,
-          components::JointVelocity().TypeId()))
-    {
-      _ecm.CreateComponent(joint, components::JointVelocity());
+      elem = elem->GetNextElement("joint_name");
     }
-
-    // Create joint force component if one doesn't exist
-    if (!_ecm.EntityHasComponentType(joint, components::JointForce().TypeId()))
+  }
+  else
+  {
+    // Create the position, velocity, and force components for the joint.
+    std::vector<Entity> childJoints = _ecm.ChildrenByComponents(
+        this->model.Entity(), components::Joint());
+    for (const Entity &joint : childJoints)
     {
-      _ecm.CreateComponent(joint, components::JointForce());
+      this->CreateComponents(_ecm, joint);
     }
   }
 }
 
+//////////////////////////////////////////////////
+void JointStatePublisher::CreateComponents(EntityComponentManager &_ecm,
+    gazebo::Entity _joint)
+{
+  this->joints.push_back(_joint);
```

Comment:
Should this check if the joint has already been added? Otherwise, if a user accidentally adds a joint twice using `joint_name`, it will be published twice.

Reply:
Changed `vector` to `set`, added a warning message, and updated the test to check for this case. 36b7b3e5f83179be42e8b67d9cd2ecb71dd81046